### PR TITLE
feat: Add support for "deno test --fail-fast=N"

### PR DESCRIPTION
**Pull request by getspooky**
Allow to specify upon n number of failing test for fail-fast flag please see #11124 discussion
By default fail-fast will stop on first error.

### CODE REVIEW

File: cli/flags.rs [hunk 1]
```diff
@@ -1001,8 +1001,16 @@ fn test_subcommand<'a, 'b>() -> App<'a, 'b> {
...
+        .help("Stop after N errors")
+        .min_values(0)
+        .required(false)
+        .takes_value(true)
+        .require_equals(true)
+        .value_name("N")
+        .validator(|val: String| match val.parse::<usize>() {
+          Ok(_) => Ok(()),
+          Err(_) => Err("fail-fast should be a number".to_string()),
+        }),
```
Comment (bartlomieju) [inline cli/flags.rs:1013]: It should also validate than `N > 0`

**Review by bartlomieju — CHANGES_REQUESTED** [review]

**Review by bartlomieju — CHANGES_REQUESTED** [review]
Also @getspooky please add an integration test to `cli/tests/integration/test_tests.rs` that check that test execution is stopped after 2 failed tests

File: cli/flags.rs (hunk 1 above)
Comment (getspooky) [reply cli/flags.rs]: fixed thanks.

File: cli/flags.rs [hunk 2]
```diff
@@ -1001,8 +1001,23 @@ fn test_subcommand<'a, 'b>() -> App<'a, 'b> {
...
+        .min_values(0)
+        .required(false)
+        .takes_value(true)
+        .require_equals(true)
+        .value_name("N")
+        .validator(|val: String| match val.parse::<usize>() {
+          Ok(val) => {
+            if val == 0 {
+              return Err(
+                "fail-fast should be an number superior to 0".to_string(),
```
Comment (bartlomieju) [inline cli/flags.rs:1014]:
```suggestion
                "fail-fast should be an number greater than 0".to_string(),
```

File: cli/tests/integration/test_tests.rs [hunk 3]
```diff
@@ -19,6 +19,21 @@ fn no_color() {
...
+    "test --fail-fast=2 test/fail_fast.ts",
+    None,
+    Some(vec![("NO_COLOR".to_owned(), "true".to_owned())]),
+    false,
+  );
+  // ANSI escape codes should be stripped.
+  assert!(out.contains("test test 1 ... FAILED"));
+  assert!(out.contains("test test 2 ... FAILED"));
+  assert!(out.contains("test result: FAILED. 0 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out"));
+}
```
Comment (bartlomieju) [inline cli/tests/integration/test_tests.rs:35]: Please use `itest!` macro and add another file for testing - depending on already existing files is bad as there's a need to change you will have to change multiple test results

**Review by bartlomieju — COMMENTED** [review]
This PR looks to me, sans 2 nitpicks below. @caspervonb can you take a look as well?

**Comment by caspervonb** [comment]
LGTM provided the test is sorted out @bartlomieju 👍

File: cli/tools/test_runner.rs [hunk 4]
```diff
@@ -540,15 +541,18 @@ pub async fn run_tests(
...
           }
           _ => {}
         }
 
         reporter.visit_event(event);
 
-        if has_error && fail_fast {
-          break;
+        if let Some(x) = fail_fast {
+          if failed >= x {
```
Comment (caspervonb) [inline cli/tools/test_runner.rs:553]: Actually, this fails on the Nth failure not after N failures as per the help text.

Comment (getspooky) [reply cli/tools/test_runner.rs:553]: so, we change the help text by 
```
Stop after Nth failure
```

Comment (bartlomieju) [reply cli/tools/test_runner.rs:553]: `Stop after N errors` would be best

Comment (getspooky) [reply cli/tools/test_runner.rs:553]: Yes , I share the same opinion 👍🏼

**Comment by bartlomieju** [comment]
Please apply this patch (copy the following code into a file called `fail_fast.patch` and then `git apply fail_fast.patch`)

```
diff --git a/cli/flags.rs b/cli/flags.rs
index df285c94e..a0f9eb2c4 100644
--- a/cli/flags.rs
+++ b/cli/flags.rs
@@ -1001,7 +1001,7 @@ fn test_subcommand<'a, 'b>() -> App<'a, 'b> {
       Arg::with_name("fail-fast")
         .long("fail-fast")
         .alias("failfast")
-        .help("Stop after N errors")
+        .help("Stop after N errors. Defaults to stopping after first failure.")
         .min_values(0)
         .required(false)
         .takes_value(true)
diff --git a/cli/tests/integration/test_tests.rs b/cli/tests/integration/test_tests.rs
index 2cf6feaf1..c029b154d 100644
--- a/cli/tests/integration/test_tests.rs
+++ b/cli/tests/integration/test_tests.rs
@@ -19,21 +19,6 @@ fn no_color() {
   assert!(out.contains("test result: FAILED. 1 passed; 1 failed; 1 ignored; 0 measured; 0 filtered out"));
 }
 
-#[test]
-fn fail_fast_with_value() {
-  let (out, _) = util::run_and_collect_output(
-    false,
-    "test --fail-fast=2 test/fail_fast.ts",
-    None,
-    Some(vec![("NO_COLOR".to_owned(), "true".to_owned())]),
-    false,
-  );
-  // ANSI escape codes should be stripped.
-  assert!(out.contains("test test 1 ... FAILED"));
-  assert!(out.contains("test test 2 ... FAILED"));
-  assert!(out.contains("test result: FAILED. 0 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out"));
-}
-
 itest!(pass {
   args: "test test/pass.ts",
   exit_code: 0,
@@ -76,12 +61,6 @@ itest!(fail_fast {
   output: "test/fail_fast.out",
 });
 
-itest!(fail_fast_with_val {
-  args: "test --fail-fast=2 test/fail_fast.ts",
-  exit_code: 1,
-  output: "test/fail_fast_with_val.out",
-});
-
 itest!(only {
   args: "test test/only.ts",
   exit_code: 1,
diff --git a/cli/tests/test/fail_fast_with_val.out b/cli/tests/test/fail_fast_with_val.out
index fe2ab807a..a902e4c7b 100644
--- a/cli/tests/test/fail_fast_with_val.out
+++ b/cli/tests/test/fail_fast_with_val.out
@@ -1,5 +1,5 @@
-Check [WILDCARD]/test/fail_fast.ts
-running 10 tests from [WILDCARD]/test/fail_fast.ts
+[WILDCARD]
+running 10 tests from [WILDCARD]/test/fail_fast_with_val.ts
 test test 1 ... FAILED ([WILDCARD])
 test test 2 ... FAILED ([WILDCARD])
 
@@ -7,17 +7,17 @@ failures:
 
 test 1
 Error
-    at [WILDCARD]/test/fail_fast.ts:2:9
+    at [WILDCARD]/test/fail_fast_with_val.ts:2:9
     at [WILDCARD]
 
 test 2
 Error
-    at [WILDCARD]/test/fail_fast.ts:5:9
+    at [WILDCARD]/test/fail_fast_with_val.ts:5:9
     at [WILDCARD]
 
 failures:
 
-    test 1
-    test 2
+	test 1
+	test 2
 
 test result: FAILED. 0 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out ([WILDCARD])
```

**Review by bartlomieju — APPROVED** [review]
LGTM, thanks @getspooky

**Pull request merged**